library/alloc/src/string.rs

```diff
@@ -1235,6 +1235,10 @@ impl String {
         let mut del_bytes = 0;
         let mut idx = 0;
 
+        unsafe {
+            self.vec.set_len(0);
+        }
+
         while idx < len {
             let ch = unsafe { self.get_unchecked(idx..len).chars().next().unwrap() };
```

Comment:
@RalfJung this was brought to my attention by @DJMcNab (the author of [`retain_more`](https://github.com/DJMcNab/retain_more)), as somewhat suspicious - it's calling `get_unchecked` on a `&str` obtained via auto-deref **which is zero-length now**. So we more or less expect this needs to go through a raw pointer directly or something similar.

Reply:
Good catch! I guess another alternative would be using a `DropGuard` for setting the length at the end of the loop. This way it's not necessary to set the length to 0 anymore because its `drop` implementation would called even in case of panic.

Reply:
I've opened #82554, let me know what you think about it